src/client/visualizations/bar-chart/bar-chart.tsx

```diff
@@ -156,242 +196,336 @@ export class BarChart extends BaseVisualization<BarChartState> {
     }
   }
 
-  onMouseEnter(measure: Measure, hoverValue: PlywoodValue, e: MouseEvent) {
-    this.setState({
-      hoverValue: { value: hoverValue },
-      hoverMeasure: measure
-    });
+  onMouseEnter(hoverInfo: BubbleInfo) {
+    this.setState({hoverInfo});
   }
 
-  onMouseLeave(measure: Measure, e: MouseEvent) {
-    const { hoverMeasure } = this.state;
-    if (hoverMeasure === measure) {
-      this.setState({
-        hoverValue: null,
-        hoverMeasure: null
-      });
+  onMouseLeave(targetHoverInfo: BubbleInfo) {
+    const { hoverInfo } = this.state;
+    if (hoverInfo && targetHoverInfo.measure === hoverInfo.measure) {
+      this.setState({hoverInfo: null});
     }
   }
 
-  onClick(measure: Measure, datum: Datum, e: MouseEvent) {
+  onClick(measure: Measure, dataPath: Datum[], splitIndex: number, selectionInfo: BubbleInfo, e: MouseEvent) {
     const { essence, clicker } = this.props;
     const { splits, dataSource } = essence;
 
-    const dimension = splits.get(0).getDimension(dataSource.dimensions);
+    const dimension = splits.get(splitIndex).getDimension(dataSource.dimensions);
 
-    var rowHighlight = getFilterFromDatum(splits, datum, dimension.name);
+    var rowHighlight = getFilterFromDatum(splits, dataPath, dataSource);
 
     if (essence.highlightOn(BarChart.id, measure.name)) {
       if (rowHighlight.equals(essence.highlight.delta)) {
         clicker.dropHighlight();
+        this.setState({selectionInfo: null});
         return;
       }
     }
 
+    this.setState({selectionInfo});
     clicker.changeHighlight(BarChart.id, measure.name, rowHighlight);
   }
 
-  renderChartBubble(dataset: Dataset, measure: Measure, chartIndex: number, containerStage: Stage, chartStage: Stage, extentY: number[], scaleY: any): JSX.Element {
-    const { essence, clicker, openRawDataModal } = this.props;
-    const { scrollTop, hoverValue, hoverMeasure, scaleX } = this.state;
+  getYExtent(measure: Measure, data: Datum[]): number[] {
```

Comment:
I think it is convention (from d3) that the data comes as the first argument and then accessor as the second